lebedigital/openbis/README.md

```diff
@@ -0,0 +1,160 @@
+)
+```
+
+> Note: You do not have to define all of those attributes, only the name, type, metadata and the location of the Sample / Object in the Database (space, project, collection) are necessary
+
+
+#### a) Define the ExpStep
+```python
+from lebedigital.openbis.expstep import ExpStep
+
+concrete_experiment = ExpStep()
+
+# The names, types, metadata, etc. are the same as defined before in examples or creating the Sample / Object Type
+
+concrete_experiment.name = 'UCT_Compression_Experiment_17_10_2022'
+concrete_experiment.type = 'EXPERIMENTAL_STEP_UCT'
+concrete_experiment.metadata = {'$name': 'Concrete_Experiment_1', 'date': '17:10:2022'}
+
+# define the location in the datastore
+
+concrete_experiment.space = '<Your Username>' # your preassigned personal space
+concrete_experiment.project = 'Unilateral Conrete Compression'
+concrete_experiment.collection = '<Your Username>/Unilateral Conrete Compression/Concrete_Compression_Testseries01'
+```
+
+#### b) Upload the ExpStep to the Database
+```python
+concrete_experiment.upload_expstep(o) # o = connected Interbis instance
+```
+#### Good to know:
+* To upload a sample / object of a given type that type has to already exist in the datastore (See 2. Create a new Sample / Object Type).
+* The name of the sample / object has to be the same as the `$name` parameter in the metadata dictionary in order to be able to find the sample / object efficiently again.
+* You can also load an existing sample / object into an ExpStep object using the `load_sample()` method of ExpStep.
+* Every time you want to upload some Sample / Object to the Database a Type Checker will run in order to make sure your ExpStep has been defined properly. The Type checker only checks whether the Metadata have the correct type and does not check for having a dataset.
+* You can look up which metadata the Experimental Step Type will accept by either getting the Excel import template running the Interbis `get_metadata_import_template()` method or get the properties of an Experimental Step Type by running the Interbis `get_sample_type_properties` method.
+
+> Return of `o.get_metadata_import_template(EXPERIMENTAL_STEP_UCT)` as an Excel Sheet or Pandas Dataframe (o = connected Interbis instance)
+
+|     | Param     | Label     | Description            | Value |
+|-----|-----------|-----------|------------------------|-------|
+| 0   | $name     | Name      | Name                   |       |
+| 1   | date      | Date      | Date of the experiment |       |
+| 2   | property1 | my_label1 | my_description1        |       |
+| 3   | property2 | my_label2 | my_description2        |       |
+| 4   | property3 | my_label3 | my_description3        |       |
+
+### 4. Upload corresponding Datasets.
+
+```python
+concrete_experiment.upload_dataset(
+    o, # A connected Interbis object
+    props={
+        '$name': 'UCT_Compression_Experiment_17_10_2022_RAW_DATA',
```

Comment:
I'm not fully sure how I would load a data set corresponding to a specific exp step, how would I define e.g. that for any compression test I have a data set load_displacement.csv that I would like to upload (and potentially also retrieve). This load_displacement file would be something like a metadata scheme (any of the tests should have such a file, no matter what the name of the file is). Or would I then have to make sure that these files are always renamed to load_displacement.csv?

Reply:
If I understand you correctly then you would need to make sure the file load_displacement.csv is always uploaded with the exp step yourself, there is no option that I know of for "forcing" the existence of a specific Dataset in Openbis. A Sample has to "exist" before you upload a Dataset to it and "uplaoding" samples is a separate step from uploading datasets. There would be no problem with uploading the file itself as Openbis will accept any file as a "Dataset". 

A solution I can think of would be to implement some sort of flag system for samples which would mark the samples which don't have the Dataset uploaded. Another would be to derive the sample upload function which would also take the load_displacement.csv file and would only upload the ExpStep if the file is given.

Also I would change the name of load_displacement.csv to contain something distinct from the sample in order to make searching for it easier,.

Reply:
The question is, how we make sure that specific files are really labelled correctly. @aradermacher mentioned that there is a specific tag in openBIS to describe the uploaded file (such that I can upload e.g. five datasets (raw_load_displacement_file, a processed file with some filtering, a figure and some files where each column is stored separately). I could then query for the raw_load_displacement file (without enforcing the name to be standardized) by using these tags (not sure of this is the correct openBIS terminology) or search for all expSteps that provide this file.

Reply:
There are predefined Dataset Types in OpenBIS like Raw Data, Processed Data, etc. and we can use those to label the Data we upload, which is useful but there is nothing stopping me from uploading a dataset that is called "Processed_Data_1" as Raw Data. A naming convention would be enforceable if you derived the Class and changed the upload function to check the Dataset names too.

What I meant with changing the name of the file is that it makes searching easier with the name only.

You could still just add the parameters for Dataset-Type = Raw Data and then download only the Dataset you want.

Reply:
I'm thinking about how to work with external departments together. For us, it is quite easy to define a naming convention and rename all the files according to schema (that we have internally defined in our pyhton_upload scripts). However, the idea is to include others, in particular the ones who are actually performing the tests. They would use the upload functionality within the website, thus we would somehow have to make sure that we can access the files in a predefined way. I guess for most of the tests. there is only a single raw data file, but e.g. the Fraunhofer guys in the last PMD meeting proposed to store column data (as processed data) per column, such that they have one time.csv, one disp1.csv, disp2.csv. If we do that (or something similar), it would actually be nice to be able to extend the datatypes to processed data time, processed_data_disp1 or processed_data_visualization. Is that possible?